### PR TITLE
feat: Ability to add optional location for each serviceRef when creating zone.

### PR DESCRIPTION
**Pull request by Ak-sky**
### Description

Added the ability to add optional location for each serviceRef when creating zone.

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

In this release-
- `cbr-service-profile` module is now modified to take optional serviceRef location when creating the zone.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by Ak-sky** [comment]
Hi @SirSpidey, please review once.

**Review by maheshwarishikha — DISMISSED** [review]

File: modules/cbr-service-profile/main.tf [hunk 1]
```diff
@@ -11,10 +11,12 @@ data "ibm_iam_account_settings" "iam_account_settings" {
...
   # tflint-ignore: terraform_unused_declarations
   validate_zone_inputs = ((length(var.zone_vpc_crn_list) == 0) && (length(var.zone_service_ref_list) == 0)) ? tobool("Error: Provide a valid zone vpc and/or service references") : true
-  # tflint-ignore: terraform_unused_declarations
-  validate_location_and_service_name = (length(setintersection(["directlink", "globalcatalog-collection", "iam-groups", "user-management"], var.zone_service_ref_list)) > 0 && var.location != null) ? tobool("Error: The services 'directlink', 'globalcatalog-collection', 'iam-groups' and 'user-management' does not support location") : true
-
 
+  # tflint-ignore: terraform_unused_declarations
+  validate_location_and_service_name = [
+    for item in ["directlink", "globalcatalog-collection", "iam-groups", "user-management"] :
+    contains(keys(var.zone_service_ref_list), item) ? length(var.zone_service_ref_list[item].serviceRef_location) == 0 ? true : tobool("Error: The services 'directlink', 'globalcatalog-collection', 'iam-groups' and 'user-management' does not support location") : true
```
Comment (SirSpidey) [inline modules/cbr-service-profile/main.tf:18]:
```suggestion
    contains(keys(var.zone_service_ref_list), item) ? length(var.zone_service_ref_list[item].serviceRef_location) == 0 ? true : tobool("Error: The services 'directlink', 'globalcatalog-collection', 'iam-groups' and 'user-management' do not support location") : true
```

File: modules/cbr-service-profile/variables.tf [hunk 2]
```diff
@@ -29,15 +33,8 @@ variable "zone_service_ref_list" {
         "apprapp", "compliance", "event-notifications", "logdna", "logdnaat",
       "cloudantnosqldb", "globalcatalog-collection", "sysdig-monitor", "sysdig-secure", "toolchain"], service_ref)
     ])
-    error_message = "Provide a valid service reference for zone creation"
+    error_message = "Provide a valid target service name that is supported by context-based restrictions"
```
Comment (SirSpidey) [inline modules/cbr-service-profile/variables.tf:36]:
```suggestion
    error_message = "Provide a valid target service name that is supported by context-based restrictions."
```

File: examples/multi-service-profile/variables.tf [hunk 3]
```diff
@@ -33,11 +27,23 @@ variable "resource_tags" {
 }
 
 variable "zone_service_ref_list" {
-  type        = list(string)
-  default     = ["cloud-object-storage", "server-protect"]
-  description = "(List) Service reference for the zone creation"
+  type = map(object({
+    serviceRef_location = optional(list(string), [])
+  }))
+  description = "Provide a valid service reference for zone creation with optional service reference location, if empty map or []/null is passed for serviceRef_location then serviceRef zone is not scoped to any location"
```
Comment (SirSpidey) [inline examples/multi-service-profile/variables.tf:33]: Does this mean, 
- Provide a valid service reference for zone creation and optionally include a service reference location.
or
- Provide a valid service reference for zone creation that has a service reference location.
or something else.

Does the next part mean,

If the map is empty, or if `null` is specified for `serviceRef_location`, the zone of the service reference zone will not be scoped to any location.

And do we need to explain why that means if it's not scoped?

**Review by SirSpidey — CHANGES_REQUESTED** [review]
I have questions about `zone_service_ref_list`. The description needs full sentences (looks like at least 2). After we set this description, apply it to all the vars in this PR.

The `zone_service_ref_list` in other variables.tf files have different descriptions. Do they need to be updated to the same?

File: examples/multi-service-profile/variables.tf (hunk 3 above)
Comment (Ak-sky) [reply examples/multi-service-profile/variables.tf:33]: > Does this mean,
> 
> * Provide a valid service reference for zone creation and optionally include a service reference location.
>   or
> * Provide a valid service reference for zone creation that has a service reference location.
>   or something else.
> 
> Does the next part mean,
> 
> If the map is empty, or if `null` is specified for `serviceRef_location`, the service reference zone will not be scoped to any location.
> 
> And do we need to explain why that means if it's not scoped?

  - `serviceRef_location` is an optional value, if not included or empty map or set to null then created service reference CBR zone will not be scoped to any location.

**Comment by Ak-sky** [comment]
> I have questions about `zone_service_ref_list`. The description needs full sentences (looks like at least 2). After we set this description, apply it to all the vars in this PR.
> 
> The `zone_service_ref_list` in other variables.tf files have different descriptions. Do they need to be updated to the same?

For now this is for `cbr-service-profile` module, still have to work on fscloud module, will update the same.

File: examples/multi-service-profile/variables.tf (hunk 3 above)
Comment (SirSpidey) [reply examples/multi-service-profile/variables.tf:33]: OK. So does this make sense? 

> Provide a location where context-based restriction zones are created. If no value is specified for `serviceRef_location`, the zones are not scoped to any location.

And if so, all the descriptions for this variable should be the same, unless there's something different in how the code uses the variable.

**Comment by SirSpidey** [comment]
> > I have questions about `zone_service_ref_list`. The description needs full sentences (looks like at least 2). After we set this description, apply it to all the vars in this PR.
> > The `zone_service_ref_list` in other variables.tf files have different descriptions. Do they need to be updated to the same?
> 
> For now this is for `cbr-service-profile` module, still have to work on fscloud module, will update the same.

I commented with an updated description. See if that makes sense.

> Provide a location where context-based restriction zones are created. If no value is specified for `serviceRef_location`, the zones are not scoped to any location.

And all the descriptions for this variable should be the same, unless there's something different in how the code uses the variable.

File: examples/multi-service-profile/variables.tf (hunk 3 above)
Comment (Ak-sky) [reply examples/multi-service-profile/variables.tf]: Yes, I have modified a little.

Comment (Ak-sky) [reply examples/multi-service-profile/variables.tf]: @SirSpidey please review again.

File: examples/multi-service-profile/variables.tf [hunk 4]
```diff
@@ -33,11 +27,23 @@ variable "resource_tags" {
 }
 
 variable "zone_service_ref_list" {
-  type        = list(string)
-  default     = ["cloud-object-storage", "server-protect"]
-  description = "(List) Service reference for the zone creation"
+  type = map(object({
+    serviceRef_location = optional(list(string), [])
+  }))
+  description = "Provide a valid service reference with location where context-based restriction zones are created. If no value is specified for serviceRef_location, the zones are not scoped to any location."
```
Comment (SirSpidey) [inline examples/multi-service-profile/variables.tf:33]:
```suggestion
  description = "Provide a valid service reference with the location where the context-based restriction zones are created. If no value is specified for `serviceRef_location`, the zones are not scoped to any location."
```

**Review by SirSpidey — CHANGES_REQUESTED** [review]
few more updates

**Review by SirSpidey — APPROVED** [review]
👍

**Pull request merged**